geoips/pydantic/products.py

```diff
@@ -0,0 +1,231 @@
+# # # This source code is protected under the license referenced at
+# # # https://github.com/NRLMMD-GEOIPS.
+
+"""Order-based procflow Product building Models."""
+
+# Python Standard Libraries
+from typing import Any, Dict, List
+
+# Third-Party Libraries
+from pydantic import BaseModel, Field, model_validator, ConfigDict
+
+# GeoIPS imports
+from geoips import interfaces
+from geoips.pydantic.bases import PluginModel
+
+
+def get_plugin_names(plugin_type: str) -> List[str]:
+    """Retrieve a list of valid plugin names for a given plugin type.
+
+    Args
+    ----
+    plugin_type
+        The type of plugin
+
+    Returns
+    -------
+        list: A list of plugin names for a valid plugin type
+
+    Raises
+    ------
+        AttributeError: If the plugin type is invalid
+
+    """
+    interface_name = plugin_type + "s"
+
+    try:
+        interface = getattr(interfaces, interface_name)
+    except AttributeError:
+        raise AttributeError(f"{plugin_type} is not valid plugin type")
+    return [plugin.name for plugin in interface.get_plugins() or []]
+
+
+def get_plugin_types() -> List[str]:
+    """Retrieve a list of plugin types from available interfaces.
+
+    Returns
+    -------
+    list of str
+        list of possible unique plugin types
+    """
+    interface = []
+    for ifs in interfaces.list_available_interfaces().values():
+        interface.extend(ifs)
+    return list(set(plugin_types[:-1] for plugin_types in interface))
+
+
+class OutputFormatterArgumentsModel(BaseModel):
+    """output_formatters."""
+
+    pass
+
+
+class FilenameFormatterArgumentsModel(BaseModel):
+    """Validate FilenameFormatter arguments."""
+
+    pass
+
+
+class AlgorithmArgumentsModel(BaseModel):
+    """Validate Algorithm arguments."""
+
+    pass
+
+
+class InterpolatorArgumentsModel(BaseModel):
+    """Validate Interpolator arguments."""
+
+    pass
+
+
+class ReaderArgumentsModel(BaseModel):
+    """Validate Reader step arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    area_def: str = Field(None, description="Area definition identifier.")
+    chans: List[str] = Field(None, description="List of channels to process.")
+    metadata_only: bool = Field(False, description="Flag for metadata-only processing.")
+    self_register: bool = Field(False, description="Flag for self-registration.")
+
+
+class ProductStepDefinitionModel(BaseModel):
+    """Validate step definition : name, arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    type: str = Field(..., description="plugin type")
+    name: str = Field(..., description="plugin name")
+    arguments: Dict[str, Any] = Field(default_factory=dict, description="step args")
+
+    @model_validator(mode="before")
+    def _validate_plugin_name(cls, values: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
I think you can do these validations without a model validator and write them explicitly into name/type